program-analysis/echidna/Exercise-X.md

```diff
@@ -0,0 +1,442 @@
+  
+   ```
+   function player_balance_is_equal_to_initial_supply() public {
+        // pre-conditions
+        uint256 currentTime = block.timestamp;
+        if (currentTime < naughtCoin.timeLock()) {
+         // post-conditions
+            assert(naughtCoin.balanceOf(player) == naughtCoin.INITIAL_SUPPLY());
+        }
+    }
+   ```
+
+</details>
+
+<details>
+  <summary>3rd property</summary>
+
+   The token transfer should fail if the current `block.timestamp < timelock`.
+
+   For this property we need to create a second user. We will try to transfer tokens from `player` to `bob`.
+
+   Add `bob` to your contract as `address bob;` and initialize his address in the constructor: `bob = address(0x123456)` to some random value.
+
+   For this property we also need to add an additional pre-condition. We need to check player's and bob's balance before transferring tokens, to have something to compare to in the post-conditions.
+
+   ```
+   function transfer_should_fail_before_timelock_period(uint256 amount) public {
+        // pre-conditions
+        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+        uint256 currentTime = block.timestamp;
+        if (currentTime < naughtCoin.timeLock()) {
+            // actions
+            naughtCoin.transfer(bob, amount);
+        }
+        // post-conditions
+        assert(
+            naughtCoin.balanceOf(player) == playerBalanceBefore &&
+                naughtCoin.balanceOf(bob) == bobBalanceBefore
+        );
+    }
+   ```
+
+Run Echidna and check the corpus. Is this a good property?
+
+</details>
+
+<details>
+  <summary>Improving the 3rd property</summary>
+
+   If you look at the corpus:
+
+   ```
+ 34 | r   |     function transfer_should_fail_before_timelock_period() public {
+ 35 |     |         // pre-conditions
+ 36 | r   |         uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+ 37 | r   |         uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+ 38 | r   |         uint256 currentTime = block.timestamp;
+ 39 | r   |         if (currentTime < naughtCoin.timeLock()) {
+ 40 |     |             // actions
+ 41 | r   |             naughtCoin.transfer(bob, 100);
+ 42 |     |         }
+ 43 |     |         // post-conditions
+ 44 |     |         assert(
+ 45 |     |             naughtCoin.balanceOf(player) == playerBalanceBefore &&
+ 46 |     |                 naughtCoin.balanceOf(bob) == bobBalanceBefore
+ 47 |     |         );
+ 48 |     |     }
+ 49 |     | }
+   ```
+
+   You can see that the transfer reverted as expected. But... Our post-conditions weren't checked. If you have read the [ERC20 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md), you will know that the `transfer` function returns a status `boolean`. In the case of success, we can emit a special `AssertionFailed` event that won't stop the execution of the function. This way, our post-conditions will be checked.
```

Comment:
What is your opinion on using `try`/`catch` instead of relying on the return value? OZ tokens either revert or return `true`:
 ```
if (currentTime < naughtCoin.timeLock()) {
         // actions
         try naughtCoin.transfer(bob,amount) { 
              // call succeeded, this is not ok
              assert(false);
         } catch {
              // call reverted
         }
}
```

Reply:
Thank you so much for your feedback, @glarregay-tob! As I described in my [blog post](https://wizzardhat.com/ethernaut-level-15-naughtcoin/), I initially went for the `try`/`catch`. They are much more intuitive and overall way better than return value checks. I switched to value checks for a couple of reasons:
- I couldn't get Echidna to work with `try`/`catch`. The property was either always passing or always reverting
- That's what Justin did in the [Intro to AMM's invariants workshop](https://www.youtube.com/watch?v=n0RaKKVTGvA), and I thought that it might be a better practice
- I must admit that in this particular case, the `try` / `catch` with an empty catch block, as explained in the [How to write good properties](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md) section is a little bit confusing to me. 

I assumed that the OZ standard returns boolean values, but as you pointed out, this is wrong. The `transfer` and `transferFrom` functions return a boolean value, but the internal `_transfer` function reverts on failure. 

Thanks again for pointing that out; I will also fix this in my article. 

I will re-write the properties with `try`/`catch` and update the exercise.

Reply:
I think it might be helpful to leave this section as it is and create a new follow-up section with Justin's setup using an external `Setup` contract and re-write properties using `try`/`catch` to showcase that both options are valid. What do you think, @glarregay-tob?

Reply:
It is weird that it didn't work for you. Let's go with the option that works first, and then if you want you can add the new section.
As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a `try` call by using `try contract.function(param) returns (bool success) {` if that helps. 
Maybe "How to write good properties" needs an update too :) thanks for pointing it out!

Reply:
> As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a try call by using try contract.function(param) returns (bool success) { if that helps. 

This might actually be the case. The `transfer` and `transferFrom` functions return true if they succeed. The reverting mechanism is caused by the internal function `_transfer`. 

I will test it out in the follow-up section. 

Thanks @glarregay-tob!

Reply:
I've been experimenting with the `try`/`catch`, and this is where I got stuck the last time. 

According to [writing good properties step by step](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md#enhacing-postcondition-checks), if we expect a function to revert, we should place the `assert(false)` in the first block. It means that the function call should not succeed, and when it does, we are bringing that up with the `assert(false)`. The following code results in a property violation. 

In this property, we expect that the `transfer` will fail.
```solidity
function token_transfer_should_fail_before_timelock_period(uint256 amount)
        public
    {
        amount = _between(amount, 0, token.INITIAL_SUPPLY());

        // pre-conditions
        uint256 currentTime = block.timestamp;
        uint256 playerBalanceBefore = token.balanceOf(address(player));
        uint256 bobBalanceBefore = token.balanceOf(address(bob));

        if (currentTime < token.timeLock()) {
            // action
            try
                player.proxy(
                    address(token),
                    abi.encodeWithSignature(
                        "transfer(address,uint256)",
                        address(bob),
                        amount
                    )
                )
            {
                assert(false);
            } catch {
                /* reverted */
            }

            // post-conditions
            assert(token.balanceOf(address(player)) == playerBalanceBefore);
            assert(token.balanceOf(address(bob)) == bobBalanceBefore);
        }
    }
```

And this Echidna's event sequence:
```
Call sequence:                                                                                                    
1.token_transfer_should_fail_before_timelock_period(0)                                                            
Event sequence:                                                                                                      
Panic(1)                                                                                                            
merror Revert 0x  
```

The `assert(false)` gets reverted. 

What does the `merror Revert 0x` mean?

I got this property working by applying your suggestion about adding the `returns(bool success)`:

```solidity
try
  player.proxy(
      address(token),
      abi.encodeWithSignature(
          "transfer(address,uint256)",
          address(bob),
          amount
      )
  )
returns (bool success, bytes memory returnData) {
  assert(!success);
} catch {
  /* reverted */
}
```

This, however, leaves me with an unused second return parameter, `returnData`. Should I ignore this warning, or is there a workaround? 

The compiler won't let me ignore the second returned data type. The `returns(bool success)` without the `bytes memory returnData` results in: `TypeError: Function returns 2 values, but returns clause has 1 variables.` 

Emitting this event also does not make sense as it won't be seen by Echidna in any case. 
```solidity
event Log(bytes data);

// ...
returns (bool success, bytes memory returnData) {
    emit Log(returnData);
    assert(!success);
} catch {
    /* reverted */
}
```

I might be missing something obvious.

Reply:
> What does the merror Revert 0x mean?

I believe this is caused by calling an external contract with the 0x0 address. Are you sure all the addresses are initialized?

Reply:
Hi @ggrieco-tob, Thanks for your response! I took a week off for a winter break. I will test this out as soon as I get back. Thanks 🎉